glibc: Fix _CS_PATH to include default profile

McDutchie opened the pull request:
###### Motivation for this change

The string returned by [`confstr(_CS_PATH)`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/confstr.html), which is also the output of the command [`getconf PATH`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/getconf.html), contains a default path that is guaranteed to find (at least) all POSIX standard utilities. The current value is:

    /run/current-system/sw/bin:/bin:/usr/bin

There are two problems with this.

1. `/bin` and `/usr/bin` should be removed, as NixOS package management never puts anything there. Anything dropped in `/bin` or `/usr/bin` under the name of a standard utility would likely be bogus.

2. The default profile binaries directory is not in it. This is a problem because the default NixOS installation lacks several mandatory standard POSIX utilities. The command recommended by the system's error message for a missing command to install a missing package *foo* is `nix-env -iA nixos.foo`. When this is done as root, the corresponding binaries are installed for all users in `/nix/var/nix/profiles/default/bin`.

###### Things done

`pkgs/development/libraries/glibc/fix_path_attribute_in_getconf.patch`:
- `CS_PATH`: Replace `/bin:/usr/bin` fallback by default profile path `/nix/var/nix/profiles/default/bin`.

Fixes: https://github.com/NixOS/nixpkgs/issues/65512

McDutchie commented:
Could someone rerun the `ofborg` check? Not sure what it does, but the results indicate it timed out before running anything, so that's not a problem with the PR.

McDutchie commented:
Ping @peti (who [committed the original value](https://github.com/NixOS/nixpkgs/blame/master/pkgs/development/libraries/glibc/fix_path_attribute_in_getconf.patch)).

SuperSandro2000 commented:
@ofborg eval

McDutchie commented:
Again, still not stale. See #65512

lheckemann commented:
I'm not sure if /bin and /usr/bin should be removed, due to /bin/sh and /usr/bin/env?

McDutchie commented:
Yet again, *not* stale. **COULD SOMEONE FINALLY PLEASE TAKE A LOOK AT THIS ALREADY???**

SuperSandro2000 commented:
> /bin and /usr/bin are unused in NixOS.

but not when nixpkgs is used in another Distro. I think it is save to add the new entry but not to remove the default entries.

> Yet again, _not_ stale.

It didn't get any attention in the last 6 months which is stale. That does not mean it is irrelevant.

McDutchie commented:
> > /bin and /usr/bin are unused in NixOS.
> 
> but not when nixpkgs is used in another Distro. I think it is save to add the new entry but not to remove the default entries.

Surely another distro that uses nixpkgs would use their own libc and not NixOS's version of glibc.

Besides, you cannot know for sure what the correct default directories on any other OS are.

SuperSandro2000 commented:
> Surely another distro that uses nixpkgs would use their own libc and not NixOS's version of glibc.

Nope, everything is linked against the libc in nixpkgs.

> Besides, you cannot know for sure what the correct default directories on any other OS are.

But ``/bin:/usr/bin`` is a pretty good guess which works 99% of the time and removing it from the defaults is bound to break things for people that use nixpkgs on none nixos like me.

Et7f3 commented:
Non nix-managed tools would most likely be linked agains an external libc.

Do we have a way to query os ? I see `stdenv.isCygwin`, `stdenv.isMips`, `stdenv.isSunOS` but no `stdenv.isNixOS`.

Maybe add this variable in another PR and use it so on non nixos ok we keep impure path but on nixos we can safely remove them.

shelvacu commented on pkgs/development/libraries/glibc/fix_path_attribute_in_getconf.patch:
```diff
@@ -3,4 +3,4 @@ diff -ubr glibc-2.17-orig/sysdeps/unix/confstr.h glibc-2.17/sysdeps/unix/confstr
 @@ -1 +1 @@
 -#define	CS_PATH	"/bin:/usr/bin"
-+#define	CS_PATH	"/run/current-system/sw/bin:/bin:/usr/bin"
++#define	CS_PATH	"/run/current-system/sw/bin:/nix/var/nix/profiles/default/bin"
```
```suggestion
+#define	CS_PATH	"/run/current-system/sw/bin:/nix/var/nix/profiles/default/bin:/usr/bin:/bin"
```

Per the other comments, we want to keep `/bin` and `/usr/bin` so that binaries work on linux systems other than NixOS. They do no harm while running on NixOS

McDutchie committed:
glibc: Fix _CS_PATH to include default profile

The string returned by confstr(_CS_PATH)[*1], which is also the
output of the command 'getconf PATH'[*2], contains a default path
that is guaranteed to find (at least) all POSIX standard utilities.
The current value is:

    /run/current-system/sw/bin:/bin:/usr/bin

The default profile binaries directory is not in it. This is a problem
because the default NixOS installation lacks several mandatory standard
POSIX utilities. The command recommended by the system's error message
for a missing command to install a missing package foo is 'nix-env -iA
nixos.foo'. When this is done as root, the corresponding binaries are
installed for all users in /nix/var/nix/profiles/default/bin.

This commit fixes these problems as follows:

pkgs/development/libraries/glibc/fix_path_attribute_in_getconf.patch:
- CS_PATH: Add default profile path '/nix/var/nix/profiles/default/bin'
  at the end.
- Retain /bin:/usr/bin; it is necessary for third-party sytems that use
  nixpkgs, linking against the nixpkgs version of glibc. It is retained
  *before* the profile directory to avoid overriding native system
  utilities on these systems (at least those that in fact store them in
  /bin and /usr/bin).

[*1] https://pubs.opengroup.org/onlinepubs/9799919799/functions/confstr.html
[*2] https://pubs.opengroup.org/onlinepubs/9799919799/utilities/getconf.html

Fixes: https://github.com/NixOS/nixpkgs/issues/65512

abathur commented:
Not sure what @doronbehar hoped I'd bring to this, so I'll just think out loud. (Sorry--I realize this may be frustrating at this point in this PR's life.)

1. The original motivating issue says:

    > The POSIX standard's idea with this is that shell scripts and other programs that invoke commands can set their $PATH to this system default value if they care about security and want to be sure to use only known-good system utilities and not anything a local user might have overridden in their own `$PATH`.
    
    and under bullet 2 in the proposed solution, also adds:
    
    > the default profile binaries directory should be added to `_CS_PATH` – at the end, so it won't be possible to override any system default packages already present.
    
    I think maybe this issue/PR combo just spent so long lost in the couch cushions that this bit wasn't in focus when shelvacu   suggested re-adding `/usr/bin:/bin` to the end of `_CS_PATH`, but I think I agree with your original assessment that the profile path belongs at the end?

    (I haven't convinced myself how much this matters, but the imperatives of NixOS and other Nix-using systems seem to be in tension here. The current location wouldn't override "system default packages" on NixOS, but it would on any other system. I imagine splash damage from this would be fairly small even as implemented, but reordering should address it.)

2. POSIX or not, a grab-bag of unspecified dependencies does seem to be swimming against "the Nix way". 

    (I don't see this as your problem to fix here--the existing patch has already been doing a form of this since 2013--but I do wonder if we should be doing something more opinionated to encourage more determinism. I'd probably be packaging the scripts and patching the getconf invocations out, myself.)

doronbehar reviewed (approved):
Changes LGTM and I think @abathur would approve them too :).
